man/amrestore.8
.\"
.de EX
.if t .ft C
.nf
..
.de EE
.fi
.if t .ft
..
.TH AMRESTORE 8
.SH NAME
amrestore \- extract backup images from an Amanda tape
.SH SYNOPSIS
.B amrestore
[
.B \-r
|
.B \-c
|
.B \-C
]
[
.B \-b
.I blocksize
]
[
.B \-f
.I fileno
]
[
.B \-l
.I label
]
[
.B \-p
]
[
.B \-h
]
.I tapedevice
|
.I holdingfile
[
.I hostname
[
.I diskname
[
.I datestamp
[
.I hostname
[
.I diskname
[
.I datestamp
\&...
]]]]]]
.SH DESCRIPTION
.B Amrestore
extracts backup images from the tape mounted on
.I tapedevice
or from the holding disk file
.I holdingfile
that match
.IR hostname ,
.I diskname
and
.I datestamp
patterns given on the command line.
The tape or holding file must be in a format written by the
.B amdump
or
.B amflush
program.
.LP
If
.I diskname
is not specified, all backups on the tape for the previous
.I hostname
are candidates.
If
.I datestamp
is not specified, all backups on the tape for the previous
.I hostname
and
.I diskname
are candidates.
If no
.IR hostname ,
.I diskname
or
.I datestamp
are specified, every backup on the tape is a candidate.
.LP
.I Hostname
and
.I diskname
are special expression descibe in the "HOST & DISK EXPRESSION" section
of
.IR amanda (8).
.I Datestamp
are special expression describe in the "DATESTAMP EXPRESSION" section
of
.IR amanda (8).
For example, if
.I diskname
is "rz[23]a", it would match disks
.B rz2a
and
.BR rz3a .
.LP
.I Datestamp
is useful if
.B amflush
writes multiple backup runs to a single tape.
.PP
Unless
.B \-p
is used,
candidate backup images are extracted to files
in the current directory named:
.LP
.RS
.I hostname.diskname.datestamp.dumplevel
.RE
.SH OPTIONS
.TP
.B \-b
Set the blocksize used to read the tape or holding file.
All holding files must be read with a blocksize of 32 KBytes.
.B Amrestore
should normally be able to determine the blocksize for tapes
on its own and not need this parameter.
.IP
The default is 32 KBytes.
.TP
.B \-f
Do a rewind followed by a fsf <fileno> before trying to restore an image.
.TP
.B \-l
Check if we restoring from the tape with the right
.I label
.TP
.B \-p
Pipe output.
The first matching backup image is sent to standard output,
which is normally a pipe to
.B restore
or
.BR tar ,
then
.B amrestore
quits.
It may be run again to continue selecting backups to process.
Make sure you specify the no-rewind
.I tapedevice
when doing this.
.IP
Note:
.B restore
may report "short read" errors when reading from a pipe.
Most versions of
.B restore
support a blocking factor option to let you set the read block size,
and you should set it to 2.
See the example below.
.TP
.B \-c
Compress output using the fastest method the compression program provides.
.B Amrestore
normally writes output files in a format understood by
.B restore
or
.BR tar ,
even if the backups on the tape are compressed.
With the
.B \-c
or
.B \-C
option,
.B amrestore
writes all files in compressed format,
even if the backups on the tape are not compressed.
Output file names will have a
.I .Z
or
.I .gz
extension depending on whether
.B compress
or
.B gzip
is the preferred compression program.
This option is useful when the current directory disk is small.
.TP
.B \-C
Compress output using the best method the compression program provides
(may be very CPU intensive).
See the notes above about the
.B \-c
option.
.TP
.B \-r
Raw output.
Backup images are output exactly as they are on the tape,
including the
.B amdump
headers.
Output file names will have a
.I .RAW
extension.
This option is only useful for debugging and other strange circumstances.
.TP
.B \-h
Header output.
The tape header block is output at the beginning of each file.
This is like
.B \-r
except
.B \-c
or
.B \-C
may also be used to compress the result.
.B Amrecover
uses the header to determine the restore program to use.
.LP
If a header is written (\-r or \-h),
only 32 KBytes are output regardless of the tape blocksize.
This makes the resulting image usable as a holding file.
.SH EXAMPLES
The following does an interactive restore of disk
.I rz3g
from host
.IR seine ,
to restore particular files.
Note the use of the
.B b
option to
.BR restore ,
which causes it to read in units of two 512-byte blocks (1 Kbyte)
at a time.
This helps keep it from complaining about short reads.
.LP
.RS
.EX
% amrestore \-p /dev/nrmt9 seine rz3g | restore \-ivbf 2 \-
.EE
.RE
.LP
The next example extracts all backup images for host
.IR seine .
This is the usual way to extract all data for a host after a disk crash.
.LP
.RS
.EX
% amrestore /dev/nrmt9 seine
.EE
.RE
.LP
If the backup datestamp in the above example is
.I 19910125
and
.I seine
has level 0 backups of disks
.I rz1a
and
.I rz1g
on the tape,
these files will be created in the current directory:
.LP
.RS
.EX
seine.rz1a.19910125.0
seine.rz1g.19910125.0
.EE
.RE
.LP
You may also use
.B amrestore
to extract a backup image from a holding disk
file that has not yet been flushed to tape:
.LP
.RS
.EX
% amrestore \-p /amanda/20001119/seine.rz1a.2 | restore \-ivbf 2 \-
.EE
.RE
.LP
.B Amrestore
may be used to generate a listing of images on a tape:
.LP
.RS
.EX
% mt -f /dev/nrmt9 rewind
% amrestore \-p /dev/nrmt9 no-such-host > /dev/null
.EE
.RE
.LP
This asks
.B amrestore
to find images for host
.BR no-such-host .
It will not find any entries that match, but along the way will report
each image it skips.
.SH CAVEATS
.LP
GNU tar must be used to restore files from backup images created with
the GNUTAR dumptype.
Vendor tar programs sometimes fail to read GNU tar images.
.SH AUTHOR
James da Silva <jds@cs.umd.edu>
.br
University of Maryland, College Park
.SH "SEE ALSO"
amanda(8),
amdump(8),
amflush(8),
tar(1)
restore(8)


NEWS
Changes in release 2.5.1b2

* debug file are seperated in subdir client/server/amandad with the config name.
* amandates client config option.
* gnutar_list-dir client config option.
* works on solaris 10 64bits.

Changes in release 2.5.1b1

* Klocwork defects fixed.
* Coverity defect fixed.
* work with tar-1.15.91
* Open SSL encryption support
* amfetchdump -o is replaced by -O.
* amrecover_do_fsf default to yes
* amrecover_check_label default to yes
* amcheck -w will do all test including the tape writable test.
  use amcheck -t -w if you want only the tape writable test.
* usetimestamps
* bsdudp/bsdtcp, 2 new auth.
* Unlimited number of DLE on a client with bsdtcp/rsh or ssh auth.
* docs/howto-auth.txt
* amandad_path, client_username and ssh_keys (for ssh/rsh)
* client config file (see man amanda-client.conf)
  Only use by amrecover.
* amrecover use the security-api
* amoldrecover which use the old amindexd/amidxtaped protocol
* -o option to overwrite config option
  (see in man amanda, the CONFIGURATION OVERWRITE section)
* new format of .amandahosts
* new format of xinetd.d/amanda

Changes in release 2.5.0p2

* listhost subcommand in amrecover

Changes in release 2.5.0p1

* Add the 'amtape update' command.

Changes in release 2.5.0

* Communication security/authentication: Kerberos 4/5, OpenSSH
* Data security: Symmetric/Assymetric encrytion algorithms (aesutil
  and gpg encryption),  Encryption can be done on server or client,
  Custom encryption utilities can be used.
* Compression: Ability to add custom compression tools. This is a
  really useful feature especially since it can specified for a DLE.  You
  can use different compression algorithm for images, binaries, ascii files
  and so on.
* Dump images spanning multiple media volumes - Dump images are no
  longer restricted to a single media volume (tape or vtape).  Data restoration
  can be done using amrecover and amfetchdump commands.
* Auto tape labelling - This optional feature is good for disk backups.

Change in release 2.4.5p1

* New [--days <num>] argument to 'amadmin <conf> tape'.
* New amoverview -last -num0 -togo0 options.

Change in release 2.4.5

* new displayunit global option to select the unit use to display number
  k=kilo, m=mega, g=giga, t=tera.
* new amoverview -skipmissed option.

Change in release 2.4.5b1

* holding disk disk use timestamped directory.
* autoflush flush today's dump.
* new bumppercent global option, this should improve bumping criteria,
  the bumpsize is set to a fixed value which can be adequate for small
  and large disk.
* bumpsize, bumppercent, bumpdays and bumpmult can be in a dumptype.
* calcsize support include and exclude like gnutar.
* new 'estimate' dumptype option to select estimate type:
    CLIENT: estimate by the dumping program.
    CALCSIZE: estimate by the calcsize program, a lot faster but less acurate.
    SERVER: estimate based on statistic from previous run, take second but
            can be wrong on the estimate size.

Changes in release 2.4.4p3

* new chg-iomega changer script.
* amanda will not use a tape if it's label is not in the tapelist file
  (eg. after an amrmtape).

Changes in release 2.4.4p2

* new initial_poll_delay config in chg-zd-mtx
* 'amadmin find' list disk removed from the disklist.
* amrecover can recover a disk removed from the disklist file.
* amanda works with samba3.
* new chg-disk changer script to use with the file: driver.
* Kerberos 4 support integrated and functional again.

Changes in release 2.4.4p1

* amverify do not advance to next tape at the end.
* new amflush -b and -s options, change -f behavior.
* new chg-mcutil changer.
* amrecover_changer works with chg-multi.
* default driveslot for chg-zd-mtx is now 0.
* amplot generate color postscript by default.
* amplot -b flag to generate b/w postscript.
* rait: works with file:
* print a postscript label for each tape used.

Changes in release 2.4.4

* New -a option to amcheck to always send an email.

Changes in release 2.4.4b1

* maxpromoteday: New option for a dumptype.
* New amtapetype program (replace tapetype)
* Client compile on CYGWIN
* amrestore -f <fileno> : amrestore can do fsf before reading the tape
* amrestore -l <label>  : amrestore can check the label before restoring.
* New config options:
  amrecover_do_fsf: Amrecover will use the -f flag of amrestore to position the
                    tape.
  amrecover_check_label: amrecover will pass the -l flag to amrestore.
  amrecover_changer: amrecover will use your changer.
  maxdumpsize: The maximum size of the dumps during a run,
               default to tapesize * runtapes
  taperalgo: select your algo for the taper, try 'largestfit', it can improve
             tape usage.
* amrecover allow to retry a skip a tape.
* New --days option to 'amadmin <config> balance' command.
* New --date option to amstatus
* amreport print an usage by tape.
* Newer promote algorithm that try to reduce the number of full on a
  single host in a run.
* New changer: chg-juke, chg-null and chg-rait

Changes in release 2.4.3

* Few small bug fixed
* New runtapes argument to amverify.
* New amverifyrun program.
* 2.4.3 client works with older server.

Changes in release 2.4.3b4

* new noop packet type. It allow the server to know the client features.
* exclude list with relative path is working again.
* amstatus is working again.

Changes in release 2.4.3b3

* --with-maxtapeblocksize configure options
* blocksize tapetype option
* file-pad tapetype option
* Multiple exclude in dumptype
* Option include in dumptype
* New disklist syntax:
*   hostname diskname [ diskdevice ] dumptype [ spindle [ interface ] ]
* chg-zd-mtx: Major cleanup and general overhaul.
* amrecover: new listdisk command.

Changes in release 2.4.3b2

* amrestore try to find the next chunk in the current directory.
* amrecover: You can change the tape device between each tape.
* amstatus work with amflush.

Changes in release 2.4.3b1

* tapeio.

installcheck/amrestore.pl
# Copyright (c) 2008, 2009, 2010 Zmanda, Inc.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
#
# Contact information: Zmanda Inc, 465 S. Mathilda Ave., Suite 300
# Sunnyvale, CA 94086, USA, or: http://www.zmanda.com

use Test::More tests => 39;
use strict;
use warnings;

use lib "@amperldir@";
use Installcheck;
use Installcheck::Config;
use Installcheck::Mock;
use Installcheck::Run qw(run run_get run_get_err run_err
		$stderr $stdout $diskname vtape_dir);
use Installcheck::Dumpcache;
use File::Path qw(rmtree mkpath);
use Amanda::Device qw( :constants );
use Amanda::Paths;
use Amanda::Constants;
use Amanda::Holding;
use Amanda::Config qw( :init :getconf );
use Amanda::Xfer qw( :constants );
use Amanda::MainLoop;
use Cwd;
use Data::Dumper;

my $testdir = "$Installcheck::TMP/amrestore-installcheck";
rmtree($testdir);
mkpath($testdir);

my $origdir = getcwd;

sub cleandir {
    for my $filename (<$testdir/*>) {
	unlink($filename);
    }
}

like(run_err('amrestore'),
    qr{Usage:},
    "'amrestore' gives usage message on stderr");

# put the debug messages somewhere
Amanda::Debug::dbopen("installcheck");
Installcheck::log_test_output();

Installcheck::Dumpcache::load("multi");
Installcheck::Run::load_vtape(1);
chdir($testdir);
my @filenames;
my ($orig_size, $comp_size, $comp_best_size, $raw_size, $hdr_size);

cleandir();
like(run_get_err('amrestore', "file:".vtape_dir()),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "simple amrestore from VFS device");

@filenames = sort <localhost.*>;
is(scalar @filenames, 2, "..and the restored files are present in testdir")
    or diag(join("\n", @filenames));
like($filenames[0], qr/localhost\..*\.[0-9]{14}\.0/,
    "..first filename looks correct");
like($filenames[1], qr/localhost\..*_dir\.[0-9]{14}\.0/,
    "..second filename looks correct");

# get the size of the first file for later
$orig_size = (stat($filenames[0]))[7];

cleandir();
Installcheck::Run::load_vtape(2);
like(run_get_err('amrestore', "file:".vtape_dir()),
    qr{Restoring from tape TESTCONF02 starting with file 1.
amrestore: 1: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 1 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "simple amrestore from VFS device (tape 2)");

@filenames = sort <localhost.*>;
is(scalar @filenames, 2, "..and the restored files are present in testdir")
    or diag(join("\n", @filenames));
like($filenames[0], qr/localhost\..*\.[0-9]{14}\.0/,
    "..first filename looks correct");
like($filenames[1], qr/localhost\..*_dir\.[0-9]{14}\.1/,
    "..second filename looks correct");

cleandir();
Installcheck::Run::load_vtape(1);
like(run_get_err('amrestore', "file:".vtape_dir(), "otherhost"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "amrestore with a nonexistent hostname restores nothing");

cleandir();
like(run_get_err('amrestore', "file:".vtape_dir(), "localhost", "/NOSuCH/DIR"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "amrestore with a good hostname but non-matching dir restores nothing");

@filenames = <localhost.*>;
is(scalar @filenames, 0, "..and no restored files are present in testdir")
    or diag(join("\n", @filenames));

cleandir();
like(run_get_err('amrestore', "file:".vtape_dir(), "localhost", "$diskname\$"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "amrestore with a good hostname and matching dir restores one file");

@filenames = <localhost.*>;
is(scalar @filenames, 1, "..and the file is present in testdir")
    or diag(join("\n", @filenames));

cleandir();
like(run_get_err('amrestore', "-l", "TESTCONF01", "file:".vtape_dir(), "localhost", "$diskname/dir"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "label checking OK");

cleandir();
ok(run('amrestore', "-b", "16384", "file:".vtape_dir()),
    "blocksize option accepted (although with no particular effect on a VFS device)");

cleandir();
like(run_get_err('amrestore', "-c", "file:".vtape_dir(), "localhost", "$diskname\$"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "compression works");

@filenames = <localhost.*>;
$comp_size = (stat($filenames[0]))[7];
ok($comp_size < $orig_size, "..compressed size is smaller than original");

cleandir();
like(run_get_err('amrestore', "-C", "file:".vtape_dir(), "localhost", "$diskname\$"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "best compression works");

@filenames = <localhost.*>;
$comp_best_size = (stat($filenames[0]))[7];
ok($comp_best_size < $comp_size, "..compressed best size is smaller than compressed fast");

cleandir();
like(run_get_err('amrestore', "-f", "2", "file:".vtape_dir()),
    qr{Restoring from tape TESTCONF01 starting with file 2.},
    "starting at filenum 2 gets no dumps");

ok(run_err('amrestore', "-c", "-r", "file:".vtape_dir()),
    "-c and -r conflict");

cleandir();
like(run_get_err('amrestore', "-r", "file:".vtape_dir(), "localhost", "$diskname\$"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "raw restore");

@filenames = <localhost.*>;
is(scalar @filenames, 1, "..and restored file is present in testdir")
    or diag(join("\n", @filenames));

# get the size of the file for later
$raw_size = (stat($filenames[0]))[7];

is($raw_size, $orig_size + Amanda::Holding::DISK_BLOCK_BYTES,
    "raw dump is orig_size + header size");

cleandir();
like(run_get_err('amrestore', "-h", "file:".vtape_dir(), "localhost", "$diskname\$"),
    qr{Restoring from tape TESTCONF01 starting with file 1.
amrestore: 1: skipping split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 0 comp N program .*
amrestore: 2: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E part 1/UNKNOWN lev 0 comp N program .*},
    "header (-h) restore");

@filenames = <localhost.*>;
is(scalar @filenames, 1, "..and restored file is present in testdir")
    or diag(join("\n", @filenames));

# get the size of the file for later
$hdr_size = (stat($filenames[0]))[7];

is($hdr_size, $orig_size + Amanda::Holding::DISK_BLOCK_BYTES,
    "header (-h) dump is orig_size + header size");

# find the holding files
config_init($CONFIG_INIT_EXPLICIT_NAME, "TESTCONF");
@filenames = sort(+Amanda::Holding::files());
is(scalar @filenames, 2, "two holding files found") or die("holding is not what I thought");
my $holding_filename = $filenames[0];

cleandir();
like(run_get_err('amrestore', $holding_filename),
    qr{Reading from '\Q$holding_filename\E'
FILE: date [0-9]* host localhost disk \Q$diskname\E lev 1 comp N program .*},
    "simple amrestore from holding disk");

@filenames = <localhost.*>;
is(scalar @filenames, 1, "..and restored file is present in testdir")
    or diag(join("\n", @filenames));

cleandir();
Installcheck::Run::load_vtape(2);
like(run_get_err('amrestore', "-h", "-p", "file:".vtape_dir(), "localhost", "$diskname/dir"),
    qr{Restoring from tape TESTCONF02 starting with file 1.
amrestore: 1: restoring split dumpfile: date [0-9]* host localhost disk \Q$diskname\E/dir part 1/UNKNOWN lev 1 comp N program .*},
    "piped amrestore");
# (note that amrestore does not go on to the next part; it used to, but would either skip or
# give an error for every subsequent file)

@filenames = <localhost.*>;
is(scalar @filenames, 0, "..leaves no files in current dir")
    or diag(join("\n", @filenames));

like($stdout, qr/AMANDA: /, "..and stdout contains something with a header");

####
# For DirectTCP, we write a dumpfile to disk manually with the NDMP device, and
# then use amrestore to get it.

SKIP: {
    skip "not built with ndmp and server", 5 unless
	Amanda::Util::built_with_component("ndmp") and Amanda::Util::built_with_component("server");

    my $ndmp = Installcheck::Mock::NdmpServer->new();
    my $port = $ndmp->{'port'};
    my $drive = $ndmp->{'drive'};

    # set up a header for use below
    my $hdr = Amanda::Header->new();
    $hdr->{type} = $Amanda::Header::F_SPLIT_DUMPFILE;
    $hdr->{datestamp} = "20091220000000";
    $hdr->{dumplevel} = 0;
    $hdr->{compressed} = 0;
    $hdr->{comp_suffix} = 'N';
    $hdr->{name} = "localhost";
    $hdr->{disk} = "/home";
    $hdr->{program} = "INSTALLCHECK";

    my $device_name = "ndmp:127.0.0.1:$port\@$drive";
    my $dev = Amanda::Device->new($device_name);
    ($dev->status() == $DEVICE_STATUS_SUCCESS)
	or die "creation of an ndmp device failed: " . $dev->error_or_status();

    $dev->start($ACCESS_WRITE, "TEST-17", "20091220000000")
	or die "could not start NDMP device in write mode: " . $dev->error_or_status();

    $dev->start_file($hdr),
	or die "could not start_file: " . $dev->error_or_status();

    {   # write to the file
	my $xfer = Amanda::Xfer->new([
		Amanda::Xfer::Source::Random->new(32768*40+280, 0xEEEEE),
		Amanda::Xfer::Dest::Device->new($dev, 0) ]);
	$xfer->start(make_cb(xmsg_cb => sub {
	    my ($src, $msg, $xfer) = @_;
	    if ($msg->{'type'} == $XMSG_ERROR) {
		die $msg->{'elt'} . " failed: " . $msg->{'message'};
	    } elsif ($msg->{'type'} == $XMSG_DONE) {
		Amanda::MainLoop::quit();
	    }
	}));

	Amanda::MainLoop::run();
    }

    $dev->finish()
	or die "could not finish device: " . $dev->error_or_status();
    undef $dev;

    pass("wrote a file to an NDMP device");

    cleandir();
    like(run_get_err('amrestore', $device_name),
	qr{Restoring from tape TEST-17 starting with file 1.
amrestore: 1: restoring split dumpfile: date [0-9]* host localhost disk /home part 1/UNKNOWN lev 0 comp N program .*},
	"simple amrestore from NDMP device");

    @filenames = sort <localhost.*>;
    is(scalar @filenames, 1, "..and the restored file is present in testdir")
	or diag(join("\n", @filenames));
    like($filenames[0], qr/localhost\..*\.[0-9]{14}\.0/,
	".. filename looks correct");

    # get the size of the first file for later
    is((stat($filenames[0]))[7], 32768*41, # note: last block is padded
	".. file length is correct");

    # let's just assume the contents of the file are correct, ok?
}

chdir("$testdir/..");
rmtree($testdir);
Installcheck::Run::cleanup();
